packages/myst-frontmatter/src/utils/referenceStash.ts

```diff
@@ -93,14 +114,47 @@ export function validateAndStashObject<T extends { id?: string; name?: string }>
   }
   if (!Object.keys(lookup).includes(value.id)) {
     // Handle case of new id - add stash value
-    lookup[value.id] = value as T & { id: string };
+    lookup[value.id] = value as WithId<T>;
+    optsLookup[value.id] = { id: value.id, opts, validateFn };
   } else if (isStashPlaceholder(lookup[value.id])) {
     // Handle case of existing placeholder { id: value, name: value } - replace stash value
-    lookup[value.id] = value as T & { id: string };
+    lookup[value.id] = value as WithId<T>;
+    optsLookup[value.id] = { id: value.id, opts, validateFn };
   } else if (warnOnDuplicate) {
     // Warn on duplicate id - lose new object
     validationWarning(`duplicate id for ${kind} found in frontmatter: ${value.id}`, opts);
   }
   stash[kind] = Object.values(lookup);
   return value.id;
 }
+
+/**
+ * Finalize and simplify stash content
+ *
+ * This also raises any warnings that were suppressed during
+ * validateAndStashObject. These may occur when a stashed placeholder
+ * is not replaced by another object.
+ */
+export function finalizeStash(stash: ReferenceStash): {
+  affiliations?: (Affiliation & { id: string })[];
+  contributors?: (Contributor & { id: string })[];
+  authorIds?: string[];
+} {
+  return {
+    affiliations: stash.affiliations?.map((aff) => {
+      if (isStashPlaceholder(aff)) {
+        const { opts, validateFn } = stash.affiliationsOpts?.[aff.id] ?? {};
+        if (opts && validateFn) validateFn(aff, { ...opts, suppressErrors: true });
```

Comment:
(...and here's where it comes back if the placeholder was never replaced by another object.)